src/devices/src/virtio/block/queue_handler.rs

```diff
@@ -0,0 +1,54 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+
+use event_manager::{EventOps, Events, MutEventSubscriber};
+use log::error;
+use vm_memory::GuestAddressSpace;
+use vmm_sys_util::epoll::EventSet;
+use vmm_sys_util::eventfd::EventFd;
+
+use crate::virtio::block::inorder_handler::InOrderQueueHandler;
+use crate::virtio::SingleFdSignalQueue;
+
+const IOEVENT_DATA: u32 = 0;
+
+// This object simply combines the more generic `InOrderQueueHandler` with a concrete queue
+// signalling implementation based on `EventFd`s, and then also implements `MutEventSubscriber`
+// to interact with the event manager. `ioeventfd` is the `EventFd` connected to queue
+// notifications coming from the driver.
+pub(crate) struct QueueHandler<M: GuestAddressSpace> {
+    pub inner: InOrderQueueHandler<M, SingleFdSignalQueue>,
+    pub ioeventfd: EventFd,
+}
+
+impl<M: GuestAddressSpace> MutEventSubscriber for QueueHandler<M> {
+    fn process(&mut self, events: Events, _ops: &mut EventOps) {
+        if events.event_set() != EventSet::IN {
```

Comment:
Should we use `contains` instead of `!=`?

You can receive error & hup events as well. Should we deregister the event in that case? Or register some sort of event for error handling that can be then processed by the vmm itself?

Nevertheless, we should not continue the processing when the event_set is not IN.

Reply:
I don't think additional flags can be encountered for an `EventFd` :-s I'll add some logic to deregister the fd when we get an error.

Reply:
I think events can contain at the same time EventSet::IN | EventSet::ERROR or EventSet::IN | EventSet::HUP.